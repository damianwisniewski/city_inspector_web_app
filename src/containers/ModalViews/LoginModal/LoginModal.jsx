import React, { Component } from 'react'
import { connect } from 'react-redux'

// SCSS
import './LoginModal.scss'

// Components
import { Form, Input } from '../../../components/FormComponents'
import { Button, Status, RequestStatus } from '../../../components/CommonComponents'
import { ModalRedirect } from '../../../components/ModalComponents/ModalRouting'

// Actions
import {
	sagaRequestLogin,
	resetLoginRequestStatus,
} from '../../../reduxStore/actionCreators/requestActions'

class LoginModal extends Component {
	state = {
		forgottenPass: false,
		email: '',
		password: '',
	}

	componentDidMount() {
		this.props.resetLoginRequestStatus()
	}

	componentDidUpdate(prevProps) {
		if (prevProps.isUserAuth) {
			this.props.closeModal()
		}
	}

	/**
	 * Handler for forgotten password button click
	 * Sets forgottenPass state to true and redirects to ForgottenPassModal
	 */
	handleForgottenPassword = () => {
		this.setState({ forgottenPass: true })
	}

	/**
	 * Handler for input changes.
	 * Sets their values in state
	 * @param {Event} e
	 */
	handleInputChanges = e => {
		const field = e.target.dataset.type

		this.setState({
			[field]: e.target.value,
		})
	}

	/**
	 * Sends provided credentials to saga responsible for login user
	 * @param {Event} e
	 */
	handleUserLogin = e => {
		e.preventDefault()

		const { email, password } = this.state
		this.props.sagaRequestLogin({ email, password })
	}

	render() {
		const { isUserAuth, loginRequestState, closeModal } = this.props
		const { email, password, forgottenPass } = this.state

		// Redirects to forgotten password modal
		if (forgottenPass) {
			return <ModalRedirect to='ForgottenPassModal' />
		}

		return (
			<div id='login-modal' className='login'>
				<h2 className='login__header'>Logowanie</h2>
				<p className='login__description'>Wpisz następujące dane aby się zalogować</p>
				<Form onSubmit={this.handleUserLogin}>
					<Input
						id='login-email-field'
						data-type='email'
						onChange={this.handleInputChanges}
						value={email}
						label='e-mail'
						placeholder='Wpisz tutaj swój email...'
						labelType='floating'
						type='email'
						autoComplete='username'
						pattern='.+@.+'
						required
					/>
					<Input
						id='login-password-field'
						data-type='password'
						onChange={this.handleInputChanges}
						value={password}
						label='hasło'
						placeholder='Podaj hasło...'
						labelType='floating'
						type='password'
						autoComplete='current-password'
						required
					/>
					<Button
						id='login-forgotten-pass-button'
						category='text'
						color='blue'
						onClick={this.handleForgottenPassword}
					>
						Zapomniałeś hasła?
					</Button>
					<div className='login__request-status'>
						<RequestStatus size='small' requestState={loginRequestState}>
							{isUserAuth ? (
								<Status id='login-status' type='correct' message='Jesteś zalogowany' />
							) : (
								<Status id='login-status' type='error' message='Niepoprawny email lub hasło' />
							)}
						</RequestStatus>
					</div>
					<div className='login__buttons-wrapper'>
						<Button id='login-close-button' color='white' onClick={closeModal}>
							Anuluj
						</Button>
						<Button id='login-submit-button' type='submit' color='blue'>
							Zaloguj
						</Button>
					</div>
				</Form>
			</div>
		)
	}
}

const mapStateToProps = state => ({
	isUserAuth: state.user.isUserAuth,
	loginRequestState: state.requests.loginRequestStatus,
})

export default connect(
	mapStateToProps,
	{
		sagaRequestLogin,
		resetLoginRequestStatus,
	},
)(LoginModal)
