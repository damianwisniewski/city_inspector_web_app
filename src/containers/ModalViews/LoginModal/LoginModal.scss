@import '../../../assets/styleHelpers/index';

.login {
	position: relative;

	width: 300px;
	min-height: 320px;

	font-size: 14px;
	text-align: center;

	&__request-status {
		min-width: 100%;
		height: 25px;
		margin-top: 7px;

		& p[class^='status'] {
			margin: 0;
		}
	}

	&__header {
		margin: 15px 10px;
		padding: 10px;

		color: $light_gray;
		font-size: 20px;
		font-family: $secondary_font;
		text-shadow: $text-shadow;

		background-color: $primary_blue;
		border-radius: 5px;
		box-shadow: $inset-shadow;
	}

	&__buttons-wrapper {
		width: 100%;
		margin-top: 10px;
		margin-bottom: 10px;
	}

	&__description {
		margin: 10px;

		color: $dark_gray;
		font-size: 15px;
	}
}
