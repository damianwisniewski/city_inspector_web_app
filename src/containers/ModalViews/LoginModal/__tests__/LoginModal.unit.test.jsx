import React from 'react'
import { configure, shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import toJson from 'enzyme-to-json'

import LoginModal from '../LoginModal'

configure({ adapter: new Adapter() })

describe('<LoginModal />', () => {
	let LoginModalElement
	const resetState = jest.fn()

	beforeEach(() => {
		LoginModalElement = shallow(
			<LoginModal.WrappedComponent
				resetLoginRequestStatus={resetState}
				loginRequestState='initial'
			/>,
		)
	})

	test('should call resetRequestState after mount', () => {
		expect(resetState).toBeCalled()
	})

	test('should render correctly LoginModal', () => {
		expect(toJson(LoginModalElement)).toMatchSnapshot()
	})

	test('should render correctly LoginModal with remember password for state forgottenPassword true', () => {
		LoginModalElement.setState({ forgottenPassword: true })
		expect(toJson(LoginModalElement)).toMatchSnapshot()
	})
})
