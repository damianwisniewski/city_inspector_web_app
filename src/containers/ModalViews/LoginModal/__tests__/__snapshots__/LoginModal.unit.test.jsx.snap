// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LoginModal /> should render correctly LoginModal 1`] = `
<div
  className="login"
  id="login-modal"
>
  <h2
    className="login__header"
  >
    Logowanie
  </h2>
  <p
    className="login__description"
  >
    Wpisz następujące dane aby się zalogować
  </p>
  <Form
    onSubmit={[Function]}
  >
    <Input
      autoComplete="username"
      data-type="email"
      id="login-email-field"
      inputRef={[Function]}
      label="e-mail"
      labelType="floating"
      onChange={[Function]}
      pattern=".+@.+"
      placeholder="Wpisz tutaj swój email..."
      required={true}
      type="email"
      value=""
    />
    <Input
      autoComplete="current-password"
      data-type="password"
      id="login-password-field"
      inputRef={[Function]}
      label="hasło"
      labelType="floating"
      onChange={[Function]}
      placeholder="Podaj hasło..."
      required={true}
      type="password"
      value=""
    />
    <Button
      category="text"
      className=""
      color="blue"
      id="login-forgotten-pass-button"
      onClick={[Function]}
      size="medium"
      type="button"
    >
      Zapomniałeś hasła?
    </Button>
    <div
      className="login__request-status"
    >
      <RequestStatus
        color="transparent"
        direction="horizontal"
        errorMessage="Przepraszamy coś poszło nie tak!"
        fullview={false}
        requestState="initial"
        size="small"
      >
        <Status
          id="login-status"
          message="Niepoprawny email lub hasło"
          position="horizontal"
          type="error"
        />
      </RequestStatus>
    </div>
    <div
      className="login__buttons-wrapper"
    >
      <Button
        category="button"
        className=""
        color="white"
        id="login-close-button"
        onClick={[Function]}
        size="medium"
        type="button"
      >
        Anuluj
      </Button>
      <Button
        category="button"
        className=""
        color="blue"
        id="login-submit-button"
        onClick={[Function]}
        size="medium"
        type="submit"
      >
        Zaloguj
      </Button>
    </div>
  </Form>
</div>
`;

exports[`<LoginModal /> should render correctly LoginModal with remember password for state forgottenPassword true 1`] = `
<div
  className="login"
  id="login-modal"
>
  <h2
    className="login__header"
  >
    Logowanie
  </h2>
  <p
    className="login__description"
  >
    Wpisz następujące dane aby się zalogować
  </p>
  <Form
    onSubmit={[Function]}
  >
    <Input
      autoComplete="username"
      data-type="email"
      id="login-email-field"
      inputRef={[Function]}
      label="e-mail"
      labelType="floating"
      onChange={[Function]}
      pattern=".+@.+"
      placeholder="Wpisz tutaj swój email..."
      required={true}
      type="email"
      value=""
    />
    <Input
      autoComplete="current-password"
      data-type="password"
      id="login-password-field"
      inputRef={[Function]}
      label="hasło"
      labelType="floating"
      onChange={[Function]}
      placeholder="Podaj hasło..."
      required={true}
      type="password"
      value=""
    />
    <Button
      category="text"
      className=""
      color="blue"
      id="login-forgotten-pass-button"
      onClick={[Function]}
      size="medium"
      type="button"
    >
      Zapomniałeś hasła?
    </Button>
    <div
      className="login__request-status"
    >
      <RequestStatus
        color="transparent"
        direction="horizontal"
        errorMessage="Przepraszamy coś poszło nie tak!"
        fullview={false}
        requestState="initial"
        size="small"
      >
        <Status
          id="login-status"
          message="Niepoprawny email lub hasło"
          position="horizontal"
          type="error"
        />
      </RequestStatus>
    </div>
    <div
      className="login__buttons-wrapper"
    >
      <Button
        category="button"
        className=""
        color="white"
        id="login-close-button"
        onClick={[Function]}
        size="medium"
        type="button"
      >
        Anuluj
      </Button>
      <Button
        category="button"
        className=""
        color="blue"
        id="login-submit-button"
        onClick={[Function]}
        size="medium"
        type="submit"
      >
        Zaloguj
      </Button>
    </div>
  </Form>
</div>
`;
